dockstore-language-plugin-parent/src/main/java/io/dockstore/language/MinimalLanguageInterface.java

```diff
@@ -0,0 +1,114 @@
+ * a) A logo for your language
+ * b) Launch instructions for your language (i.e. given all the desciptors for a workflow locally, how would I run a workflow on the command-line?)
+ * c) What language is your description in (Markdown, HTML, etc.)?
+ * d) A short (acronym like CWl, WDL) name for your language and a longer more descriptive name (like "Common Workflow Language")
+ */
+public interface MinimalLanguageInterface {
+
+    /**
+     * Validate a filename path that might be entered by your user (i.e. is "/Dockstore.wdl" a valid path to the "first" descriptor).
+     * This is used to distinguish between multiple workflows that may be stored in one repo.
+     *
+     * @return a pattern that can be used to validate filepaths
+     */
+    Pattern initialPathPattern();
+
+    /**
+     * Find all files that would be useful to display to a user (hopefully all files needed to launch a workflow)
+     *
+     * @param initialPath path to the intial descriptor as set by the user (e.g. /example_directory/Dockstore.cwl)
+     * @param contents    contents of the initial descriptor
+     * @param reader      get additional files and their content
+     * @return a map from absolute paths (relative to the root of the repo, e.g. "common.cwl" as opposed to "../common.cwl") to their file types and content
+     */
+    Map<String, Pair<String, GenericFileType>> indexWorkflowFiles(String initialPath, String contents, FileReader reader);
+
+    /**
+     * Given the primary descriptor and the files indexed from indexWorkflowFiles, return relevant metadata
+     * about the workflow.
+     *
+     * @param initialPath  the path to the primary descriptor
+     * @param contents     contents of the primary descriptor
+     * @param indexedFiles the set of files indexed above
+     * @return the workflow metadata that we will show to users
+     */
+    WorkflowMetadata parseWorkflowForMetadata(String initialPath, String contents, Map<String, Pair<String, GenericFileType>> indexedFiles);
+
+    /**
+     * When indexing, Dockstore will distinguish between extra files that hold things like extra code, tools, configuration (imported descriptors) and test parameter files (example parameter sets used to run a workflow)
+     */
+    enum GenericFileType { IMPORTED_DESCRIPTOR, TEST_PARAMETER_FILE, CONTAINERFILE }
+
+    /**
+     * Reads files relative to the file found via {@link #initialPathPattern()}
+     */
+    interface FileReader {
+        /**
+         * Get the contents of a file from GitHub, GitLab, etc.
+         *
+         * @param path relative to the primary descriptor, the path
+         * @return contents of the file at path
+         */
+        String readFile(String path);
+    }
+
+    /**
+     * Information that can be parsed from a workflow and would be useful to display to users
+     */
+    class WorkflowMetadata {
```

Comment:
In the existing languages, I believe the workflow metadata always comes from the version. While it doesn't hurt -- maybe some language will support workflow-level metadata, I wonder if we have this until we run into such a case...

Just raising this as something to think about...

Reply:
In this context, this is always harvesting metadata from a specific version from github, gitlab, hosted, etc. I'll rename to make this clear.